---
title: 'metagene of genes'
author: "Manfred Schmid"
output: 
  pdf_document:
    toc: true 
    toc_depth: 3
    fig_caption: true
---

`r format(Sys.time(), "%d %B, %Y; %R")`

# Setup

```{r setup, echo=TRUE}
knitr::opts_chunk$set(fig.width=12, fig.height=8, 
                      fig.path=paste0('../Figures/CLIP_genes/'), 
                      dev='pdf', 
                      echo=TRUE, warning=FALSE, message=FALSE, 
                      error=TRUE)
```

```{r load packages, echo=T, warning=F, message=F}
suppressWarnings(library('tidyverse'))
suppressWarnings(library('magrittr'))
suppressWarnings(library('knitr'))
suppressWarnings(library('RMetaTools'))
```

# load Annotation


```{r}
load('../../data/subRead_tr_annotations.RData', verbose=T)
tr_anno_tbl
```

```{r}
tr_anno_tbl %<>% 
  filter(gene_type == 'protein_coding') %>%
  dplyr::select(GeneID, exon_cnt, tr_exons_width) %>%
  mutate(class = ifelse(exon_cnt == 1, 'monoexonic', 'multiexonic'))
```

```{r}
pc_ids <- unique(tr_anno_tbl$GeneID)

length(pc_ids)
```




# Load CLIP data


## NCBP3 and EIF4A3 from CLIPdb

Data for NCBP3 (c17orf85 and EIF4A3) are from CLIPdb study, mapped to hg38

#### metagene values using deeptools
```{bash, eval = FALSE}
#!/bin/sh
##cd /home/schmidm/faststorage/CLIP/CLIPdb/scripts
##sbatch --account=thj_common --mem=4g deeptools_subReadanno_metagene.sh

. /home/schmidm/miniconda2/etc/profile.d/conda.sh
conda activate deeptools3

#these annotations are shipped with subRead
#anno="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.txt"

#sed 1d $anno | sort -k2,2 -k1,1 -k3,3n | \
#awk '{
#   if(gene_id == $1){
#     starts=starts","($3-start-1)
#     sizes=sizes","($4-$3+1)
#     end=$4
#     n+=1
#   }else{
#     if(gene_id != ""){
#       print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts
#     }
#     gene_id=$1; chr=$2; start=$3-1; end=$4; strand=$5; n=1;
#     starts="0"; sizes=($4-$3+1)
#   }
# }END{print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts}' | sort -k1,1 -k2,2n > ${anno/.txt/.bed}

bed12="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.bed"

#awk '{if($6 =="+"){print $0}}' $bed12 > ${bed12/.bed/_plus.bed}
#awk '{if($6 =="-"){print $0}}' $bed12 > ${bed12/.bed/_minus.bed}

plus_bw=$(ls /home/schmidm/faststorage/CLIP/CLIPdb/hg38_bw/*_plus_hg38.bw | awk '$1 ~ /C17orf85/ || $1 ~ /CSTF2/ || $1 ~ /EIF4A3/' | tr "\n" " ")

minus_bw=${plus_bw//_plus_hg38.bw/_minus_hg38.bw}

python ~/ms_tools/MS_Metagene_Tools/computeMatrixStranded.pyc scale-regions -Rp ${bed12/.bed/_plus.bed} -Rm ${bed12/.bed/_minus.bed} -Sp ${plus_bw} -Sm ${minus_bw} --metagene -o deeptools_subReadanno_metagene_scaled.gz -b 1000 -a 1000 --binSize=50 --regionBodyLength=1000 --numberOfProcessors=1 --quiet

```


#### load to R
```{r}
fname <- '/Volumes/GenomeDK/faststorage/CLIP/CLIPdb/scripts/deeptools_subReadanno_metagene_scaled.gz'

df <- RMetaTools::load_deeptoolsmatrix3(fname)
```

```{r}
(df %<>% 
   filter(id %in% pc_ids) %>%
    dplyr::mutate(sample_name = sub('.*\\/', '', sample_name) %>%
                         sub('_plus_hg38.bw', '', .)) %>%
  dplyr::select(id, sample_name, rel_pos, value) %>%
   dplyr::mutate(GeneID = as.character(id)) %>%
   left_join(., tr_anno_tbl))
```


#### save

```{r}
ncbp3 <- filter(df, sample_name == 'C17orf85_PARCLIP_PARalyzer')

saveRDS(ncbp3, file='../data/NCBP3_CLIP_gene_metagene.rds')
```

```{r}
eif4a3 <- filter(df, sample_name == 'EIF4A3_HITSCLIP_Piranha_001')

saveRDS(eif4a3, file='../data/EIF4A3_CLIP_gene_metagene.rds')
```



## NCBP2 from Giacometti et al

#### deeptools run
```{bash, eval = FALSE}
#!/bin/sh
##cd /home/schmidm/faststorage/CLIP/Giacometti_GSE94427/scripts
##sbatch --account=thj_common --mem=4g deeptools_subReadanno_metagene.sh

. /home/schmidm/miniconda2/etc/profile.d/conda.sh
conda activate deeptools3

#these annotations are shipped with subRead
#anno="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.txt"

#sed 1d $anno | sort -k2,2 -k1,1 -k3,3n | \
#awk '{
#   if(gene_id == $1){
#     starts=starts","($3-start-1)
#     sizes=sizes","($4-$3+1)
#     end=$4
#     n+=1
#   }else{
#     if(gene_id != ""){
#       print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts
#     }
#     gene_id=$1; chr=$2; start=$3-1; end=$4; strand=$5; n=1;
#     starts="0"; sizes=($4-$3+1)
#   }
# }END{print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts}' | sort -k1,1 -k2,2n > ${anno/.txt/.bed}

bed12="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.bed"

#awk '{if($6 =="+"){print $0}}' $bed12 > ${bed12/.bed/_plus.bed}
#awk '{if($6 =="-"){print $0}}' $bed12 > ${bed12/.bed/_minus.bed}

plus_bw=$(ls /home/schmidm/faststorage/CLIP/Giacometti_GSE94427/hg38/*_plus_hg38.bw | tr "\n" " ")
minus_bw=${plus_bw//_plus_hg38.bw/_minus_hg38.bw}

python ~/ms_tools/MS_Metagene_Tools/computeMatrixStranded.pyc scale-regions -Rp ${bed12/.bed/_plus.bed} -Rm ${bed12/.bed/_minus.bed} -Sp ${plus_bw} -Sm ${minus_bw} --metagene -o deeptools_subReadanno_metagene_scaled.gz -b 1000 -a 1000 --binSize=50 --regionBodyLength=1000 --numberOfProcessors=1 --quiet

```



```{r}
fname <- '/Volumes/GenomeDK/faststorage/CLIP/Giacometti_GSE94427/scripts/deeptools_subReadanno_metagene_scaled.gz'

df <- RMetaTools::load_deeptoolsmatrix3(fname)
```

```{r}
(cbp20 <- df %>% 
   filter(grepl('CBP20', sample_name), id %in% pc_ids) %>%
    dplyr::mutate(sample_name = sub('.*GSM......._', '', sample_name) %>%
                         sub('_norm_plus_hg38.bw', '', .)) %>%
  dplyr::select(id, sample_name, rel_pos, value) %>%
   dplyr::mutate(GeneID = as.character(id)) %>%
   left_join(., tr_anno_tbl))
```



#### average replicates
The datasets are replicates that behave nicely (not shown here), so we simply average over the 2 replicates.
```{r}
cbp20 %<>%
  mutate(sample_name = sub('_.*', '', sample_name)) %>%
  group_by(id, sample_name, rel_pos, GeneID, exon_cnt, tr_exons_width, class) %>%
  summarize(value = sum(value)/2)
```

#### save cbp20 data

```{r}
saveRDS(cbp20, file='../data/CBP20_CLIP_gene_metagene.rds')
```


```{r}
rm(df)
```

## ALY data from Viphakone et al

## deeptools run
```{bash, eval = FALSE}
#!/bin/sh
##cd /project/THJ_common/faststorage/people/MS/Yuhui/Viphakone_etal
##sbatch --account=thj_common --mem=4g deeptools_subReadanno_metagene.sh

. /home/schmidm/miniconda2/etc/profile.d/conda.sh
conda activate deeptools3

#these annotations are shipped with subRead
#anno="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.txt"

#sed 1d $anno | sort -k2,2 -k1,1 -k3,3n | \
#awk '{
#   if(gene_id == $1){
#     starts=starts","($3-start-1)
#     sizes=sizes","($4-$3+1)
#     end=$4
#     n+=1
#   }else{
#     if(gene_id != ""){
#       print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts
#     }
#     gene_id=$1; chr=$2; start=$3-1; end=$4; strand=$5; n=1;
#     starts="0"; sizes=($4-$3+1)
#   }
# }END{print chr"\t"start"\t"end"\t"gene_id"\t0\t"strand"\t"start"\t"end"\t255,0,0\t"n"\t"sizes"\t"starts}' | sort -k1,1 -k2,2n > ${anno/.txt/.bed}

bed12="/home/schmidm/ms_tools/subread-2.0.0-Linux-x86_64/annotation/hg38_RefSeq_exon.bed"

#awk '{if($6 =="+"){print $0}}' $bed12 > ${bed12/.bed/_plus.bed}
#awk '{if($6 =="-"){print $0}}' $bed12 > ${bed12/.bed/_minus.bed}

plus_bw=$(ls /home/schmidm/THJ_common/faststorage/data/Human/GEO/GSE113896/hg38/*plus*.bw | tr "\n" " ")
minus_bw=${plus_bw//_hg38_plus.bw/_hg38_minus.bw}

python ~/ms_tools/MS_Metagene_Tools/computeMatrixStranded.pyc scale-regions -Rp ${bed12/.bed/_plus.bed} -Rm ${bed12/.bed/_minus.bed} -Sp ${plus_bw} -Sm ${minus_bw} --metagene -o deeptools_subReadanno_metagene_scaled.gz -b 1000 -a 1000 --binSize=50 --regionBodyLength=1000 --numberOfProcessors=1 --quiet

```



```{r}
fname <- '/Volumes/GenomeDK/THJ_common/faststorage/people/MS/Yuhui/Viphakone_etal/deeptools_subReadanno_metagene_scaled.gz'

df <- RMetaTools::load_deeptoolsmatrix3(fname)
```

```{r}
(aly <- df %>% 
   filter(grepl('Alyref', sample_name), id %in% pc_ids) %>%
    dplyr::mutate(sample_name = sub('.*GSE113896_', '', sample_name) %>%
                         sub('-union_hg38', '', .)) %>%
  dplyr::select(id, sample_name, rel_pos, value) %>%
   dplyr::mutate(GeneID = as.character(id)) %>%
   left_join(., tr_anno_tbl))
```


#### save aly data


```{r}
saveRDS(aly, file='../data/ALY_CLIP_gene_metagene.rds')
```

```{r}
rm(df)
```


#### alternative starting point

```{r}
ncbp3 <- readRDS('../data/NCBP3_CLIP_gene_metagene.rds')

ncbp3$sample_name <- 'NCBP3'
```

```{r}
eif4a3 <- readRDS('../data/EIF4A3_CLIP_gene_metagene.rds')

eif4a3$sample_name <- 'EIF4A3'
```

```{r}
cbp20 <- readRDS('../data/CBP20_CLIP_gene_metagene.rds')
```

```{r}
aly <- readRDS('../data/ALY_CLIP_gene_metagene.rds')

aly$sample_name <- 'ALYREF'
```


#### combine

```{r}
df <- bind_rows(ncbp3, eif4a3) %>% 
  bind_rows(., cbp20) %>% 
  bind_rows(., aly)
```


# Plots

#### plot fun

```{r}
metaplot_all <- function(df) {
  df %>% 
    group_by(sample_name, rel_pos) %>%
    summarize(events=n()) %>%
    ggplot(., aes(x=rel_pos, y=events, color=sample_name)) +
    geom_line() +
    facet_wrap(~sample_name, scales='free') +
    theme_bw() +
    theme(panel.grid=element_blank())
}
```

```{r}
metaplot_perclass <- function(df) {
  df %>% 
    group_by(class, sample_name, rel_pos) %>%
    summarize(events=n()) %>%
    ggplot(., aes(x=rel_pos, y=events, color=sample_name)) +
    geom_line() +
    facet_wrap(~sample_name+class, scales='free') +
    theme_bw() +
    theme(panel.grid=element_blank())
}
```


```{r all genes}
metaplot_all(df)
```



```{r mono vs multiexonic}
metaplot_perclass(df)
```



```{r mono vs multiexonic transcripts g200nt}
df %>% 
  filter(class == 'multiexonic', tr_exons_width > 200) %>%
  metaplot_perclass
```

```{r multiexonic only transcripts g200nt}
df %>% 
  filter(class == 'multiexonic', tr_exons_width > 200) %>%
  metaplot_all
```


# sessionInfo
```{r}
sessionInfo()
```
